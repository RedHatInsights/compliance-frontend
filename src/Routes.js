import { Route, Switch, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import React from 'react';
import asyncComponent from './Utilities/asyncComponent';
import some from 'lodash/some';

/**
 * Aysnc imports of components
 *
 * https://webpack.js.org/guides/code-splitting/
 * https://reactjs.org/docs/code-splitting.html
 *
 * pros:
 *      1) code splitting
 *      2) can be used in server-side rendering
 * cons:
 *      1) nameing chunk names adds unnecessary docs to code,
 *         see the difference with DashboardMap and InventoryDeployments.
 *
 */
const CompliancePolicies = asyncComponent(() =>
    import(/* webpackChunkName: "CompliancePolicies" */ './SmartComponents/CompliancePolicies/CompliancePolicies')
);
const ComplianceSystems = asyncComponent(() =>
    import(/* webpackChunkName: "ComplianceSystems" */ './SmartComponents/ComplianceSystems/ComplianceSystems')
);
const PolicyDetails = asyncComponent(() =>
    import(/* webpackChunkName: "PolicyDetails" */ './SmartComponents/PolicyDetails/PolicyDetails')
);
const SystemDetails = asyncComponent(() =>
    import(/* webpackChunkName: "SystemDetails" */ './SmartComponents/SystemDetails/SystemDetails')
);

export const paths = {
    compliancePolicies: '/policies',
    complianceSystems: '/systems',
    complianceSystemsInventoryDetail: '/systems/:inventoryId',
    policyDetails: '/policies/:policy_id',
    policyDetailsInventoryDetail: '/policies/:policy_id/:inventoryId',
    systemDetails: '/systems/:inventoryId'
};

type Props = {
    childProps: any
};

const InsightsRoute = ({ component: Component, rootClass, ...rest }) => {
    const root = document.getElementById('root');
    root.removeAttribute('class');
    root.classList.add(`page__${rootClass}`, 'pf-c-page__main');
    root.classList.add(`page__${rootClass}`, 'pf-l-page__main');
    root.setAttribute('role', 'main');

    return (<Component { ...rest } />);
};

InsightsRoute.propTypes = {
    component: PropTypes.func,
    rootClass: PropTypes.string
};

/**
 * the Switch component changes routes depending on the path.
 *
 * Route properties:
 *      exact - path must match exactly,
 *      path - https://prod.foo.redhat.com:1337/insights/compliance/rules
 *      component - component to be rendered when a route has been chosen.
 */
export const Routes = (props: Props) => {
    const path = props.childProps.location.pathname;

    return (
        <Switch>
            <Route exact path={paths.compliancePolicies} component={CompliancePolicies} />
            <Route exact path={paths.complianceSystems} component={ComplianceSystems} />
            <Route path={paths.complianceSystemsInventoryDetail} component={SystemDetails} />
            <Route path={paths.policyDetailsInventoryDetail} component={SystemDetails} />
            <Route exact path={paths.policyDetails} component={PolicyDetails} />
            <Route path={paths.systemDetails} component={SystemDetails} />

            { /* Finally, catch all unmatched routes */ }
            <Route render={ () => some(paths, p => p === path) ? null : (<Redirect to={paths.compliancePolicies}/>) }/>
        </Switch>
    );
};
