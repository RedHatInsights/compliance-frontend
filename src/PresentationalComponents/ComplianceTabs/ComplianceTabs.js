import routerParams from '@redhat-cloud-services/frontend-components-utilities/files/RouterParams';
import propTypes from 'prop-types';
import React from 'react';
import { paths } from '../../Routes';
import { Tabs, Tab } from '@patternfly/react-core';
import invert from 'lodash/invert';

export const ComplianceTabs = (props) => {
    const { match: { path } } = props;

    const tabPaths = {
        0: paths.complianceReports,
        1: paths.compliancePolicies,
        2: paths.complianceSystems
    };

    const redirect = (_, tabIndex) => {
        props.history.push(tabPaths[tabIndex]);
    };

    let currentKey = Number(invert(tabPaths)[path]);

    const tabs = [
        <Tab title={'Reports'} key={0} eventKey={0}></Tab>,
        <Tab title={'Policies'} key={1} eventKey={1}></Tab>,
        <Tab title={'Systems'} key={2} eventKey={2}></Tab>
    ];

    return (
        <Tabs activeKey={currentKey} onSelect={redirect} aria-label="Compliance Tabs">
            { tabs }
        </Tabs>
    );
};

ComplianceTabs.propTypes = {
    history: propTypes.object,
    match: propTypes.object
};

export default routerParams(ComplianceTabs);
