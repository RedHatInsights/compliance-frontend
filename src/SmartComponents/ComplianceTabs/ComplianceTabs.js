import { routerParams } from '@red-hat-insights/insights-frontend-components';
import propTypes from 'prop-types';
import React, { Component } from 'react';
import { paths } from '../../Routes';
import { Nav, NavList, NavItem, NavVariants } from '@patternfly/react-core';

class ComplianceTabs extends Component {
    constructor(props) {
        super(props);
        this.redirect = this.redirect.bind(this);
    }
    redirect(tab) {
        this.props.history.push(tab.itemId);
    }
    render() {
        return (
            <React.Fragment>
                <Nav onSelect={this.redirect} aria-label="Compliance Tabs">
                    <NavList variant={NavVariants.horizontal}>
                        <NavItem preventDefault={true}
                            isActive={this.props.match.path === paths.compliancePolicies}
                            itemId={paths.compliancePolicies}>
                            Policies
                        </NavItem>
                        <NavItem preventDefault={true}
                            isActive={this.props.match.path === paths.complianceSystems}
                            itemId={paths.complianceSystems}>
                            Systems
                        </NavItem>
                    </NavList>
                </Nav>
            </React.Fragment>
        );
    }
}

ComplianceTabs.propTypes = {
    history: propTypes.object,
    match: propTypes.object
};

export default routerParams(ComplianceTabs);
