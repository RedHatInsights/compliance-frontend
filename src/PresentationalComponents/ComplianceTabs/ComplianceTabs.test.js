import toJson from 'enzyme-to-json';
import { paths } from '../../Routes';

import { ComplianceTabs } from './ComplianceTabs';

describe('ComplianceTabs', () => {
    const defaultProps = {
        history: {},
        match: {
            path: paths.complianceSystems
        }
    };

    it('expect to render without error', () => {
        const wrapper = shallow(
            <ComplianceTabs { ...defaultProps } />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    // TODO: Add test triggering a click on a tab to test "redirect"
});
