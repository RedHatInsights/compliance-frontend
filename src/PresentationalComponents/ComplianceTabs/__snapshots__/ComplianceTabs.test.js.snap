// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComplianceTabs expect to render without error 1`] = `
<Component
  activeKey={2}
  aria-label="Compliance Tabs"
  onSelect={[Function]}
>
  <Tab
    eventKey={0}
    key="0"
    title="Reports"
  />
  <Tab
    eventKey={1}
    key="1"
    title="Policies"
  />
  <Tab
    eventKey={2}
    key="2"
    title="Systems"
  />
</Component>
`;
